engine/minskyTensorOps.cc

```diff
@@ -276,9 +276,67 @@ namespace minsky
   {
     std::shared_ptr<ITensor> arg1, arg2;
     void computeTensor() const override {//TODO
-      throw runtime_error("inner product not yet implemented");
+		
+      if (arg1->rank()>2 || arg2->rank()>2)
+        throw runtime_error("inner product of rank>2 tensors not yet implemented");
+   		 
+      size_t stride=arg2->hypercube().dims()[0];	
+      size_t m=arg1->hypercube().dims()[0], n=arg2->hypercube().dims()[arg2->rank()-1];	 
+      double tmpSum;
+      for (size_t i=0; i< m; i++) {
+        if (arg1->rank()>1 && arg2->rank()>1)		
+          for (size_t j=0; j< n; j++)
+            {
+              tmpSum=0;
+              for (size_t k=0; k<stride; k++)  {
+				auto v1=arg1->atHCIndex(i*stride+k);  
+				auto v2=arg2->atHCIndex(k*stride+j);  
+                tmpSum+=(isnan(v1)? 0 : v1) * (isnan(v2)? 0 : v2);
+			   }
+              cachedResult[j+i*stride]=tmpSum;
+            }
+        else {
+          for (size_t k=0; k<stride; k++)  {
+			auto v1=(arg1->rank()>1? arg1->atHCIndex(i*stride+k):(*arg1)[i]);   
+			auto v2=(*arg2)[k];
+            cachedResult[i]+= (isnan(v1)? 0 : v1)*(isnan(v2)? 0 : v2);
+		  }
```

Comment:
Is there a significant performance advantage in having special case code for rank 1 arguments? Might be worth profiling.

Reply:
> Might be worth profiling to see if the extra code complexity is worth it.

I remember we discussed profiling in February, but I don't remember which one you use or recommend for optimizing the code?

Reply:
I use gperftools. Do make CPUPROFILE=1 to enable (just the link step is needed, so rather than building everything, just delete the minsky executable).

then run minsky with
env CPUPROFILE=prof.out gui-tk/minsky

The run "pprof --callgrind gui-tk/minsky prof.out >prof.cg"

Finally run "kcachegrind prof.cg"

You'll need to install the relevant packages for these tools, of course.

Reply:
Thank you :)

Reply:
On my machine pprof is google-pprof. is that correct? pprof seems to be something else, a part of a package called "tau".

I also had to replace -lprofiler in Minsky's Makefile with "LIBS+=-L/usr/lib/x86_64-linux-gnu/libprofiler" because as usual, Ubuntu has stuff in non-standard places...

So, maybe that's the wrong libprofiler.so because I only generate empty ".tmp.minsky.*.prof.out" files in the new "pprof" folder under my /home directory?

I've made all the other changes you suggested and the segfault related to destroying a non-existent boost::any pointer is gone, thank you :) I'll commit the code now.